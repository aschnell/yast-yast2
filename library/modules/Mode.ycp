/**
 * File:	modules/Mode.ycp
 * Module:	yast2
 * Summary:	Installation mode
 * Authors:	Klaus Kaempf <kkaempf@suse.de>
 * Flags:	Stable
 *
 * $Id$
 *
 * Provide installation mode information.
 * Mostly values from /etc/install.inf
 * See linuxrc documentation for detailed docs about this.
 */

{

module "Mode";

textdomain "base";

/**
 * Current mode
 */
string _mode = nil;

/**
 * Current testing mode
 */
string _test = nil;

/**
 * We do one automatic check whether _test should be set to testsuite.
 */
boolean test_autochecked = false;

/**
 * Current UI mode
 */
string _ui = "dialog";


/**
 * initialize everything from command-line of y2base
 */
global void Initialize () {
    _mode = "normal";
    _test = "none";
    integer arg_count = size(WFM::Args());
    integer arg_no = 0;
    while ( arg_no < arg_count )
    {
	// parsing for main mode
	if (WFM::Args(arg_no) == "initial"
	    || WFM::Args(arg_no) == "continue"
	    || WFM::Args(arg_no) == "firstboot")
        {
	    _mode = "installation";
        }
	// parsing for test mode
	else if (WFM::Args(arg_no) == "test" || WFM::Args(arg_no) == "demo")
	{
	    _test = "test";
	    y2warning("***** Test mode enabled *****");
	}
	else if (WFM::Args(arg_no) == "screenshots"  )
	{
	    _test = "screenshot";
	    y2warning("***** Screen shot mode enabled *****");
	}

	arg_no = arg_no + 1;
    }

    // only use the /etc/install.inf agent when file is present
    // and installation is being processed
    // FIXME remove the part below and let it be set in clients
    if (_mode == "installation"
	&& SCR::Read (.target.size, "/etc/install.inf") != -1)
    {
	boolean autoinst = SCR::Read (.etc.install_inf.AutoYaST) != nil;
        if (autoinst)
        {
	    _mode="autoinstallation";
        }

	boolean repair = SCR::Read (.etc.install_inf.Repair ) != nil;
        if (repair)
        {
	    _mode ="repair";
        }
    }
}

// main mode definitions

/**
 * Returns the current mode name. It's one of
 * "installation", "normal", "update", "repair", "autoinstallation", "autoinst_config"
 */
global string mode () {
    if (_mode == nil)
    {
	Initialize ();
    }

    return _mode;
}

global void SetMode (string new_mode) {
    if (_mode == nil)
	Initialize ();

    if (! contains (
	[ "installation", "update", "normal", "repair",
	    "autoinstallation", "autoinst_config",
	    "live_installation"
	],
	new_mode))
    {
	y2error ("Unknown mode %1", new_mode);
    }

    y2milestone ("setting mode to %1", new_mode);
    _mode = new_mode;
}

global boolean commandline();

// test mode definitions

global string testMode () {
    if (_test == nil)
	Initialize ();
    if (!test_autochecked) {
        // bnc#243624#c13: Y2ALLGLOBAL is set by yast2-testsuite/skel/runtest.sh
        if (getenv ("Y2MODETEST") != nil || getenv ("Y2ALLGLOBAL") != nil) {
	    _test = "testsuite";
	}
	test_autochecked = true;
    }

    return _test;
}

global void SetTest (string new_test_mode) {
    if (_test == nil)
	Initialize ();

    if (! contains (
	[ "none", "test", "demo", "screenshot", "testsuite", ],
	new_test_mode))
    {
	y2error ("Unknown test mode %1", new_test_mode);
    }
    _test = new_test_mode;
}

// UI mode definitions

/**
 * Returns the current UI mode.
 * It's one of "commandline", "dialog", "none"
 */
global string ui () {
    return _ui;
}

global void SetUI (string new_ui) {
    if (! contains (
	[ "commandline", "dialog", "none", ],
	new_ui))
    {
	y2error ("Unknown UI mode %1", new_ui);
    }
    _ui = new_ui;
}

// main mode wrappers

/**
 * we're doing a fresh installation
 */
global boolean installation () {
    return mode () == "installation" || mode () == "autoinstallation"
	|| mode () == "live_installation";
}

/**
 * we're doing a fresh installation from live CD/DVD
 */
global boolean live_installation () {
    return mode () == "live_installation";
}

/**
 * we're doing an update
 */
global boolean update () {
    return mode () == "update";
}

global boolean Depeche () {
    return true;
}

/**
 * normal, running system
 */
global boolean normal () {
    return mode () == "normal";
}

/**
 * start repair module
 */
global boolean repair () {
    return mode () == "repair";
}

/**
 * doing auto-installation
 */
global boolean autoinst () {
    return mode () == "autoinstallation";
}

/**
 * configuration for auto-installation, only in running system
 */
global boolean config () {
    return mode () == "autoinst_config";
}

// test mode wrappers

/**
 * Just testing.
 * See installation/Test-Scripts/doit*
 */
global boolean test () {
    return testMode () == "test" || testMode () == "screenshot"
        || testMode () == "testsuite";
}

/**
 * dump screens to /tmp. Implies @ref #demo .
 * See installation/Test-Scripts/yast2-screen-shots*
 */
global boolean screen_shot () {
    return testMode () == "screenshot";
}

/**
 * Returns whether running in testsuite.
 */
global boolean testsuite () {
    return testMode () == "testsuite";
}

// UI mode wrappers

/**
 * we're running in command line interface
 * @return true if command-line is running
 */
global boolean commandline () {
    return ui () == "commandline";
}

} // EOF
