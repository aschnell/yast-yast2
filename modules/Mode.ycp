/**
 * File:
 *	Mode.ycp
 *
 * Module:
 *	Mode
 *
 * Summary:
 *	provide installation mode information
 *	Mostly value from /etc/install.inf
 *	See linuxrc documentation for detailed docs about this.
 *
 * $Id$
 *
 * Author:
 *	Klaus Kaempf <kkaempf@suse.de>
 *
 */

{
    module "Mode";

    /**
     * we're doing a fresh installation
     */
    global boolean installation = true;

    /**
     * we're doing an update
     */
    global boolean update	= false;

    /**
     * /var/adm/mount has installation data mounted
     */
    global boolean sourcemounted= false;

    /**
     * how we were booted (the type of the installation medium)
     * /etc/install.inf: Bootmode
     */
    global string boot		= "cd";

    /**
     * just testing
     */
    global boolean test		= false;

    /**
     * dry-run of dialogues
     */
    global boolean demo		= false;	// includes test_mode if set

    /**
     * dump screens to /tmp
     */
    global boolean screen_shot	= false;	// includes demo_mode if set

    /**
     * starting installation in inst-sys system
     */
    global boolean initial	= false;

    /**
     * continuing installation in target system
     */
    global boolean cont		= false;

    /**
     * hard or softboot when starting installed system ?
     */
    global boolean hardBoot	= false;

    /**
     * normal, running system
     */
    global boolean normal	= false;

    /**
     * do auto-installation
     */
    global boolean autoinst	= false;

    /**
     * configuration for auto-installation, only in running system
     */
    global boolean config	= false;

    /**
     * Preparing live-eval cd installation
     */
    global boolean live_eval	= false;

    /**
     * this is a beta-version
     */
    global boolean beta		= false;	// this is a beta version

    /**
     * no resources/packages for X11
     */
    global boolean text_fallback	= false;

    /**
     * somehow, no X11 was started
     */
    global boolean no_x11		= false;	// no x11 or not enough memory for qt

    /**
     * is set when called due to a hardware change
     */
    global boolean reprobe		= false;

    /**
     * we're running in textmode (-> UI::GetDisplayInfo())
     */
    global boolean text		= false;

    /**
     * we're running in manual mode (manual installation choosen in linuxrc)
     */
    global boolean manual		= false;

    /**
     * running via serial console
     */
    global boolean serial_console	= false;

    /**
     * Data from /etc/install.inf (linuxrc related)
     */
    global map installMap		= $[];

    /**
     * language set in install.inf
     */
    global string language		= "en_US";

    /**
     * braille mode ?
     */
    global boolean braille		= false;

    /**
     * vnc mode ?
     */
    global boolean vnc			= false;

    /**
     * install map containing data for the online update
     */
    global map youInstallMap		= $[];

    /**
     * save important data to installmap.ycp
     */
    global define void Save ()
    ``{
	installMap["updatemode"] = update;
	installMap["hardBoot"] = hardBoot;
	SCR::Write (.target.ycp, [Installation::yast2dir + "/installmap.ycp", 0600], installMap);
    }

    //---------------------------------------------------------------
    * constructor

    global define Mode () ``{

	/*
	 * test command line arguments for test mode and continue mode
	 */

	integer arg_count = size(WFM::Args());
	integer arg_no = 0;

	while ( arg_no < arg_count )
	{
	    y2debug("option #%1: %2", arg_no, WFM::Args(arg_no) );

	    if	    (WFM::Args(arg_no) == "test"	   )
	    {
		test = true;
		y2warning("***** Test mode enabled *****");
	    }
	    else if (WFM::Args(arg_no) == "initial"	    ) initial	= true;
	    else if (WFM::Args(arg_no) == "continue"	    ) cont	= true;
	    else if (WFM::Args(arg_no) == "demo"	    )
	    {
		demo = true;
		y2warning("***** Demo mode enabled *****");
	    }
	    else if (WFM::Args(arg_no) == "screenshots"  )
	    {
		screen_shot = true;
		y2warning("***** Screen shot mode enabled *****");
	    }
	    else if (WFM::Args(arg_no) == "text_fallback") text_fallback   = true;
	    else if (WFM::Args(arg_no) == "no_x11"       ) no_x11	      = true;
	    else if (WFM::Args(arg_no) == "reprobe"      ) reprobe	      = true;
	    else
	    {
		y2milestone ("skipping unknown option %1", WFM::Args(arg_no) );
	    }

	    arg_no = arg_no + 1;
	}


	if (!initial
	    && !cont
	    && !autoinst
	    && !config)
	{
	    normal = true;
	}

	// screen_shot ==> demo
	if ( screen_shot )	demo = true;

	// demo ==> test
	if ( demo )	test = true;


	// get setup data from linuxrc
	// check setup/descr/info for CD type

	if (initial)
	{
	    if ( !test )
	    {
		installMap = SCR::Read(.etc.install_inf);
	    }
	    SCR::UnmountAgent (.etc.install_inf);
	    y2milestone("Reading /etc/install.inf: initial installMap: %1", installMap);
	}
	else
	{
	    installMap = SCR::Read (.target.ycp, ["/var/lib/YaST2/installmap.ycp", $[]]);
	    y2milestone("Reading /var/lib/YaST2/installmap.ycp: installMap: %1", installMap);
	    /**
	     * restore update only during installation/update, leave
	     * it false in 'normal' mode.
	     */
	    if (!Mode::normal)
	    {
		update = installMap["updatemode"]:false;
	    }
	    hardBoot = installMap["hardBoot"]:false;
	}

	language	= installMap["language"]:"";
	live_eval	= installMap["demo"]:"0" == "1";
	text		= installMap["textmode"]:"0" == "1";
	manual		= installMap["manual"]:"" == "1";
	serial_console	= installMap["console"]:"" != "";
 	braille		= installMap["braille"]:"" != "";
	vnc		= installMap["runvnc"]:"" == "1";
	boot		= installMap["instmode"]:"cd";
	sourcemounted	= installMap["sourcemounted"]:"" == "1";
	// the value is a string
	autoinst	= installMap["autoinstall"]:"" != "";
    }
}
