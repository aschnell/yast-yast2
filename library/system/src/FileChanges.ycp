/**
 * File:	modules/FileChanges.ycp
 * Module:	yast2
 * Summary:	Detect if a configuratil file was changed
 * Authors:	Jiri Srain <jsrain@suse.cz>
 *
 * Support routines for detecting changes of configuration files being done
 * externally (not by YaST) to prevent the changes from being lost because
 * of YaST not handling the configuration files correctly (eg. removing
 * comments in some cases, changing order of options,...)
 *
 * Warns user if such change is detected.
 *
 * Usage:
 * Before reading the configuration file:
 *   call boolean CheckFiles (list<string>) with all files. If any of them
 *   is detected to be changed, YaST asks a popup for you.
 *   alternatively use boolean FileChanged (string) for each file (does not
 *   ask any question, immediatelly returns status of the file
 *
 * After writing the configuraiton file:
 *   call void StoreFileCheckSum (string) for each file to store recent
 *   checksum. YaST will use this checksum next time checking.
 *
 */

{

module "FileChanges";

textdomain "base";

import "Mode";
import "Popup";
import "Directory";
import "Label";

string data_file = "/var/lib/YaST2/file_checksums.ycp";

map<string,string> file_checksums = $[];

/**
 * Read the data file containing file checksums
 */
void ReadSettings () {
    if ((integer)SCR::Read (.target.size, data_file) <= 0)
    {
	file_checksums = $[];
	return;
    }
    file_checksums = (map<string,string>)SCR::Read (.target.ycp, data_file);
    if (file_checksums == nil)
	file_checksums = $[];
}

/**
 * Write the data file containing checksums
 */
void WriteSettings () {
    SCR::Write (.target.ycp, data_file, file_checksums);
}

/**
 * Compute the checksum of a file
 * @param file string the file to compute checksum of
 * @return string the checksum
 */
string ComputeFileChecksum (string file) {
    // See also FileUtils::MD5sum()
    string cmd = sformat ("/usr/bin/md5sum %1", file);
    map out = (map)SCR::Execute (.target.bash_output, cmd);
    // note: it also contains file name, but since it is only to be compared
    // it does not matter
    string sum = out["stdout"]:"";
    return sum;
}

/**
 * Check if file was modified compared to the one distributed
 * with the RPM package
 * @param file string the file to check
 * @return boolean true of was changed
 */
boolean FileChangedFromPackage (string file) {
    // queryformat: no trailing newline!
    string cmd = sformat ("/bin/rpm -qf %1 --qf %%{NAME}-%%{VERSION}-%%{RELEASE}", file);
    map out = (map)SCR::Execute (.target.bash_output, cmd);
    string package = out["stdout"]:"";
    y2milestone ("Package owning %1: %2", file, package);
    if (package == "" || out["exit"]:-1 != 0)
	return false;
    cmd = sformat ("rpm -V %1 |grep ' %2$'", package, file);
    out = (map)SCR::Execute (.target.bash_output, cmd);
    string changes = out["stdout"]:"";
    y2milestone ("File possibly changed: %1", changes);
    list<string> lines = splitstring (changes, "\n");
    boolean changed = false;
    foreach (string line, lines, {
	if (regexpmatch (line, "^S"))
	    changed = true;
	if (regexpmatch (line, "^..5"))
	    changed = true;
	if (regexpmatch (line, "^.......T"))
	    changed = true;
    });
    return changed;
}

/**
 * Check if a file was modified externally (without YaST)
 * @param file string boolean the file to check
 * @return boolean true if was changed externally
 */
global boolean FileChanged (string file) {
    // when generating AutoYaST configuration, they are not written back
    if (Mode::config ())
	return false;
    ReadSettings ();
    boolean ret = false;
    if (haskey (file_checksums, file))
    {
	y2milestone ("Comparing file %1 to stored checksum", file);
	string sum = ComputeFileChecksum (file);
	ret = ! (sum == file_checksums[file]:"");
    }
    else
    {
	y2milestone ("Comparing file %1 to RPM database", file);
	ret = FileChangedFromPackage (file);
    }
    y2milestone ("File differs: %1", ret);
    return ret;
}

/**
 * Store checksum of a file to the store
 * @param file string filename to compute and store
 */
global void StoreFileCheckSum (string file) {
    ReadSettings ();
    string sum = ComputeFileChecksum (file);
    file_checksums[file] = sum;
    WriteSettings ();
}

/**
 * Check files if any of them were changed
 * Issue a question whether to continue if some were chaned
 * @param files a list of files to check
 * @return boolean true if either none was changed or user agreed
 *  to continue
 */
global boolean CheckFiles (list<string> files) {
    files = filter (string f, files, { return FileChanged (f); });
    if (size (files) > 0)
    {
	// Continue/Cancel question, %1 is a file name
	string msg = _("File %1 has been changed manually.
YaST might lose some of the changes.
");
	if (size (files) > 1)
	    // Continue/Cancel question, %1 is a coma separated list of file names
	    msg = _("Files %1 have been changed manually.
YaST might lose some of the changes");
	msg = sformat (msg, mergestring (files, ", "));
	string popup_file = "/filechecks_non_verbose";
	if ($[] == SCR::Read (.target.stat, Directory::vardir + popup_file))
	 {
	  term content = `VBox(
		`Label(msg),
		`Left(`CheckBox(`id(`disable), _("Do not show this message anymore"))),
		`ButtonBox (
		    `PushButton(`id(`ok), `opt (`okButton), Label::ContinueButton()),
		    `PushButton(`id(`cancel), `opt (`cancelButton), Label::CancelButton())
		)
	    );
	  UI::OpenDialog(content);
	  UI::SetFocus(`ok);
	  any ret=UI::UserInput();
	  y2milestone("ret = %1", ret);
	  if (ret==`ok && (boolean)UI::QueryWidget(`disable, `Value)){
	   y2milestone("Disabled checksum popups");
	   SCR::Write ( .target.string, Directory::vardir + popup_file, "");
	   }
	  UI::CloseDialog();
	  if (ret==`ok)return true;
		else return false;
	 }
//			return Popup::ContinueCancel (msg);
		else return true;
    }
    return true;
}

}
