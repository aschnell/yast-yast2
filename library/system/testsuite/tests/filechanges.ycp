/**
 * File:
 *      filechanges.ycp
 *
 * Module:
 *      Base
 *
 * Summary:
 *      testsuite for filechanges module
 *
 * Authors:
 *      Jiri Srain <jsrain@suse.cz>
 *
 * $Id: kernel.ycp 26253 2005-11-22 14:59:29Z lslezak $
 *
 */

// testedfiles: Kernel.ycp Testsuite.ycp

{
    include "testsuite.ycp";

    // minimal "don't care" SCR data to work with the constructor
    map READ = $[
	"target" : $[
	    "string" : "",
	    "tmpdir" : "/tmp",
	    "size" : 5,
	    "ycp" : $[
	    ],
	],
    ];
    map WRITE = $[];
    list EXECL = [ $[
	"target" : $[
	    "bash_output" : $[
		"exit" : 0,
		"stdout" : "ntp.rpm",
		"stderr" : ""
	    ],
	],
    ],
    $[
	"target" : $[
	    "bash_output" : $[
		"exit" : 0,
		"stdout" : "S.?....T    /etc/ntp.conf
..?.....  c /var/lib/ntp/etc/ntp.conf.iburst",
		"stderr" : ""
	    ],
	],
    ],
    ];

    TESTSUITE_INIT ([READ, $[], $[]], 0);

    import "FileChanges";
    import "Mode";

    Mode::SetTest ("testsuite");
    TEST (``(FileChanges::FileChanged ("/etc/ntp.conf")), [READ, WRITE, EXECL], nil);
    READ["target", "ycp", "/etc/ntp.conf"] = "incorrect checksum";
    map EXEC = $[
	"target" : $[
	    "bash_output" : $[
		"exit" : "0",
		"stdout" : "f210720e1362615ac0ecc544b35abb73  /etc/ntp.conf",
	    ],
	],
    ];

    TEST (``(FileChanges::FileChanged ("/etc/ntp.conf")), [READ, WRITE, EXEC], nil);
    READ["target", "ycp", "/etc/ntp.conf"] = "f210720e1362615ac0ecc544b35abb73  /etc/ntp.conf";
    TEST (``(FileChanges::FileChanged ("/etc/ntp.conf")), [READ, WRITE, EXEC], nil);
}
