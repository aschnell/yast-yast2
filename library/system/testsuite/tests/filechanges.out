Return	true
Return	true
Return	false
